### PR TITLE
MOD-6540: Support EMPTY indexing for TEXT fields

### PR DESCRIPTION
**Pull request by raz-mon**
**Describe the changes in the pull request**

Currently we discard empty strings for TEXT fields in our index, i.e., we don't index them and don't allow querying for them as well.

This PR adds the support of indexing and querying for such empty strings (`''`) for TEXT fields.

This is done by adding the `ISEMPTY` keyword in the index creation for the desired TEXT field, as such:
```c
FT.CREATE idx SCHEMA my_field TEXT ISEMPTY
```

Then, querying for such empty values is done via the `isempty(@my_field)` syntax in the query, as such:
```c
FT.SEARCH idx "isempty(@myfield) ... "
```

This syntax is compatible with all other query constructs, just as any other first-class citizen.

Some cleaning and commenting was done as well, 

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: src/spec.c [hunk 1]
```diff
@@ -810,6 +864,19 @@ static int parseVectorField(IndexSpec *sp, StrongRef sp_ref, FieldSpec *fs, Args
   }
 }
 
+static int parseGeometryField(IndexSpec *sp, FieldSpec *fs, ArgsCursor *ac, QueryError *status) {
```
Comment (nafraf) [inline src/spec.c:867]: This function always return 1, is it correct?

**Review by nafraf — COMMENTED** [review]
Good. Some comments

File: tests/pytests/test_empty.py [hunk 2]
```diff
@@ -0,0 +1,655 @@
+from common import *
+import json
+
+EMPTY_RESULT = [0]
+
+def testEmptyValidations():
+    """Validates the edge-cases of the `ISEMPTY` field option"""
+
+    env = Env(moduleArgs="DEFAULT_DIALECT 2")
```
Comment (nafraf) [inline tests/pytests/test_empty.py:9]: Should we test for DIALECT 3 an 4?

File: src/query.c [hunk 3]
```diff
@@ -806,10 +812,6 @@ static IndexIterator *Query_EvalFuzzyNode(QueryEvalCtx *q, QueryNode *qn) {
 }
 
 static IndexIterator *Query_EvalPhraseNode(QueryEvalCtx *q, QueryNode *qn) {
```
[the remark this answers is not in the document]
Comment (raz-mon) [reply src/query.c:814]: I think so yes, but I don't want to clean the whole code on this PR..

File: src/spec.c (hunk 1 above)
Comment (raz-mon) [reply src/spec.c:867]: Yes
I agree that we don't need it.

File: src/document.c [hunk 4]
```diff
@@ -449,8 +445,12 @@ FIELD_PREPROCESSOR(fulltextPreprocessor) {
...
 
       Token tok = {0};
-      uint32_t newTokPos;
-      while (0 != (newTokPos = aCtx->tokenizer->Next(aCtx->tokenizer, &tok))) {
+      while (0 != aCtx->tokenizer->Next(aCtx->tokenizer, &tok)) {
+        if ((!strcmp(tok.tok, "")) && !FieldSpec_IndexesEmpty(fs)) {
+          // Skip empty values if the field should not index them
+          // Empty tokens are returned only if the original value was empty
+          continue;
+        }
```
Comment (oshadmi) [inline src/document.c:453]: Can move out of the loop?
```suggestion
      bool isIndexEmpty = FieldSpec_IndexesEmpty(fs);
      while (0 != aCtx->tokenizer->Next(aCtx->tokenizer, &tok)) {
        if (!isIndexEmpty && (!strlen(tok.tok))) {
          // Skip empty values if the field should not index them
          // Empty tokens are returned only if the original value was empty
          continue;
        }
```

File: src/document.c [hunk 5]
```diff
@@ -695,7 +689,7 @@ FIELD_PREPROCESSOR(geoPreprocessor) {
     case FLD_VAR_T_BLOB_ARRAY:
     case FLD_VAR_T_NUM:
     case FLD_VAR_T_GEOMETRY:
-      RS_LOG_ASSERT(0, "Oops");
+      RS_LOG_ASSERT(0, "Unsupported field type GEOMETRIY for GEO index");
```
Comment (oshadmi) [inline src/document.c:692]: Could also be `FLD_VAR_T_NUM` or `FLD_VAR_T_BLOB_ARRAY`
```suggestion
      RS_LOG_ASSERT(0, "Unsupported field type for GEO index");
```

File: src/document.c (hunk 4 above)
Comment (raz-mon) [reply src/document.c:453]: Good idea in general.
But the `isIndexEmpyt` test must come after the test for the empty string, otherwise we will just not index any token in the case of non-empty indexing field :)

File: tests/pytests/test_short_read.py [hunk 6]
```diff
@@ -27,7 +27,7 @@
     'short-reads/redisearch_2.2.0_rejson_2.0.0.rdb.zip',
     'short-reads/redisearch_2.8.0.rdb.zip',
     'short-reads/redisearch_2.8.4.rdb.zip',
-    'short-reads/redisearch_2.8.12_rejson_2.0.0.rdb.zip',
+    'short-reads/redisearch_2.10.3.rdb.zip',
```
Comment (nafraf) [inline tests/pytests/test_short_read.py:30]: Why did you remove the 2.8.12 rdb test? is it not relevant?

Comment (raz-mon) [reply tests/pytests/test_short_read.py:30]: The new rdb contains both empty TAG and TEXT, and the name was a bit misleading since we did not merge this into 2.8.

**Review by nafraf — APPROVED** [review]

File: src/numeric_index.c [hunk 7]
```diff
@@ -655,34 +655,7 @@ struct indexIterator *NewNumericFilterIterator(RedisSearchCtx *ctx, const Numeri
...
-      t = NewNumericRangeTree();
-      RedisModule_ModuleTypeSetValue((*idxKey), NumericIndexType, t);
-    } else {
-      t = RedisModule_ModuleTypeGetValue(*idxKey);
-    }
-  } else {
-    t = openNumericKeysDict(ctx->spec, keyName, 1);
-  }
-  return t;
+  return openNumericKeysDict(ctx->spec, keyName, 1);
```
Comment (oshadmi) [inline src/numeric_index.c:658]: The rest was dead-code being cleaned up, right?

File: src/query.c [hunk 8]
```diff
@@ -471,6 +471,14 @@ IndexIterator *Query_EvalTokenNode(QueryEvalCtx *q, QueryNode *qn) {
   // we can just use the optimized score index
   int isSingleWord = q->numTokens == 1 && q->opts->fieldmask == RS_FIELDMASK_ALL;
 
+  // If this is an empty value search, and the field does not index empty values
+  // throw an error
+  if (qn->tn.nen == NON_EXIST_EMPTY && !(FieldSpec_IndexesEmpty(IndexSpec_GetFieldByBit(q->sctx->spec, qn->opts.fieldMask)))) {
+    QueryError_SetErrorFmt(q->status, QUERY_ENOINDEX, "Field `%s` does not index empty values",
```
Comment (oshadmi) [inline src/query.c:477]: Consider a positive message potentially assisting the user, such as "Field `%s` should enable `ISEMPTY` in the index SCHEMA in order to support empty values"

File: src/query.c [hunk 9]
```diff
@@ -1315,6 +1317,11 @@ static IndexIterator *Query_EvalTagNode(QueryEvalCtx *q, QueryNode *qn) {
   IndexIterator *ret = NULL;
 
   if (!idx) {
+    if (qn->tag.nen == NON_EXIST_EMPTY && !FieldSpec_IndexesEmpty(fs)) {
+      QueryError_SetErrorFmt(q->status, QUERY_ENOINDEX,
```
Comment (oshadmi) [inline src/query.c:1321]: Consider a positive message

File: src/query.c [hunk 10]
```diff
@@ -1723,7 +1730,7 @@ static sds QueryNode_DumpSds(sds s, const IndexSpec *spec, const QueryNode *qs,
       s = sdscat(s, "}");
       break;
     case QN_TOKEN:
-      s = sdscatprintf(s, "%s%s", (char *)qs->tn.str, qs->tn.expanded ? "(expanded)" : "");
+      s = sdscatprintf(s, "%s%s", strcmp(qs->tn.str, "") == 0 ? "<ISEMPTY>" : qs->tn.str, qs->tn.expanded ? "(expanded)" : "");
```
Comment (oshadmi) [inline src/query.c:1733]: Can use `SPEC_INDEXEMPTY_STR`?

File: src/rlookup.h [hunk 11]
```diff
@@ -72,17 +72,14 @@ typedef struct RLookupKey {
   uint16_t svidx;
 
   /**
-   * Can be F_SVSRC which means the target array is a sorting vector, or
-   * F_OUTPUT which means that the t
+   * Can be F_SVSRC which means the target array is a sorting vector)
```
Comment (oshadmi) [inline src/rlookup.h:75]: There is another remaining occurrence of `F_OUTPUT`  :up:

File: src/numeric_index.c (hunk 7 above)
Comment (raz-mon) [reply src/numeric_index.c:658]: Yes

File: src/spec.c [hunk 12]
```diff
@@ -171,6 +171,17 @@ const char *IndexSpec_GetFieldNameByBit(const IndexSpec *sp, t_fieldMask id) {
   return NULL;
 }
 
+// Get the field spec by the field mask.
+const FieldSpec *IndexSpec_GetFieldByBit(const IndexSpec *sp, t_fieldMask id) {
```
Comment (oshadmi) [inline src/spec.c:175]: Only for TEXT? If so, name is a bit misleading

Comment (raz-mon) [reply src/spec.c:175]: But I agree in general.

File: src/spec.c [hunk 13]
```diff
@@ -748,10 +711,101 @@ static int parseVectorField_flat(FieldSpec *fs, VecSimParams *params, ArgsCursor
...
+
+  return rc;
+}
+
 static int parseVectorField(IndexSpec *sp, StrongRef sp_ref, FieldSpec *fs, ArgsCursor *ac, QueryError *status) {
   // this is a vector field
   // init default type, size, distance metric and algorithm
 
+  fs->types |= INDEXFLD_T_VECTOR;
+  sp->flags |= Index_HasVecSim;
```
[the remark this answers is not in the document]
Comment (oshadmi) [reply src/spec.c:807]: Oh OK - moved from the caller

File: tests/pytests/test_empty.py [hunk 14]
```diff
@@ -0,0 +1,655 @@
...
+            'h1', ['t', '', 'text', 'hello'],
+            'h9', ['t', ','],
+            'h5', ['t', ', bar'],
+            'h7', ['t', 'bat,'],
+            'h6', ['t', 'bat, '],
+            'h8', ['t', 'bat, , bat2']
+        ]
+        cmd_assert(env, cmd, expected)
+
+        # ------------------------ Priority vs. Intersection -----------------------
```
Comment (oshadmi) [inline tests/pytests/test_empty.py:232]: Priority meaning parser precedence? (`isempty` precedes intersection)

File: tests/pytests/test_empty.py [hunk 15]
```diff
@@ -0,0 +1,655 @@
...
+
+    # Searching for emptiness of a non-existing field should return an error
+    obj = {
+        'obj': {}
+    }
+    objs = json.dumps(obj, separators=(',', ':'))
+    env.expect('JSON.SET', 'j', '$', objs).equal('OK')
+    cmd = f'FT.SEARCH jidx isempty(@obj)'.split(' ')
+    expected = EMPTY_RESULT
+    env.expect(*cmd).error().contains('Syntax error: Field not found')
```
Comment (oshadmi) [inline tests/pytests/test_empty.py:456]: Can be removed since covered by `testEmptyValidations`?

File: tests/pytests/test_empty.py [hunk 16]
```diff
@@ -0,0 +1,655 @@
...
+    env.flush()
+
+    # Test that when we index many docs, we find the wanted portion of them upon
+    # empty value indexing
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 't', 'TAG', 'ISEMPTY').ok()
+    n_docs = 1000
+    for i in range(n_docs):
+        conn.execute_command('HSET', f'h{i}', 't', '' if i % 2 == 0 else f'{i}')
+    res = env.cmd('FT.SEARCH', 'idx', 'isempty(@t)', 'LIMIT', '0', '0')
+    env.assertEqual(int(res[0]), 500)
```
Comment (oshadmi) [inline tests/pytests/test_empty.py:483]: Consider also testing the negation returns 500 results.

File: tests/pytests/test_empty.py [hunk 17]
```diff
@@ -0,0 +1,655 @@
...
+    """Tests the indexing and querying of empty TEXT (field type) values"""
+
+    env = Env(moduleArgs="DEFAULT_DIALECT 2")
+    conn = getConnectionByEnv(env)
+
+    def testEmptyTextHash(idx):
+        """Tests the indexing and querying of empty values for a TEXT field of a
+        hash index
+        Extensive tests are added here, specifically to the query part, due to
+        the addition of the `isempty` function syntax added to the parser.
```
Comment (oshadmi) [inline tests/pytests/test_empty.py:495]: Consider removing this (would become less relevant as time passes)

File: tests/pytests/test_empty.py [hunk 18]
```diff
@@ -0,0 +1,655 @@
...
+
+        # ------------------------------- Prefix -------------------------------
+        cmd = f'FT.SEARCH {idx} @t:*pty'.split(' ')
+        expected = EMPTY_RESULT
+        cmd_assert(env, cmd, expected)
+
+        # ------------------------------- Suffix -------------------------------
+        cmd = f'FT.SEARCH {idx} @t:__em*'.split(' ')
+        expected = EMPTY_RESULT
+        cmd_assert(env, cmd, expected)
```
Comment (oshadmi) [inline tests/pytests/test_empty.py:553]: Irrelevant now when using `isempty(@t)` instead of `__empty`?

File: tests/pytests/test_empty.py [hunk 19]
```diff
@@ -0,0 +1,655 @@
...
+        # We shouldn't get the document when searching for a prefix of "__empty"
+        cmd = f'FT.SEARCH {idx} @t:{{*pty}}'.split(' ')
+        expected = EMPTY_RESULT
+        cmd_assert(env, cmd, expected)
+
+        # ------------------------------- Suffix -------------------------------
+        # We shouldn't get the document when searching for a suffix of "__empty"
+        cmd = f'FT.SEARCH {idx} @t:{{__em*}}'.split(' ')
+        expected = EMPTY_RESULT
+        cmd_assert(env, cmd, expected)
```
Comment (oshadmi) [inline tests/pytests/test_empty.py:137]: Irrelevant now when using `isempty(@t)` instead of `__empty`?

File: tests/pytests/test_empty.py [hunk 20]
```diff
@@ -0,0 +1,655 @@
...
+
+    # Test that we get in error in case of a user tries to use "isempty(@field)"
+    # when `field` does not index empty values.
+    env.expect('FT.SEARCH', 'idx', 'isempty(@tag)').error().contains('does not index empty values')
+    env.expect('FT.SEARCH', 'idx', 'isempty(@text)').error().contains('does not index empty values')
+    # Empty search on a non-existing field
+    env.expect('FT.SEARCH', 'idx', 'isempty(@non_existing)').error().contains('Field not found')
+
+
+def EmptyJSONTest(env, idx):
```
Comment (oshadmi) [inline tests/pytests/test_empty.py:24]: Consider renaming to something like EmptyJSONTagTest?
Or make it a an inner function as well? (like `testEmptyTextHash`)

What about empty text with JSON?

(there's `testEmptyTextHash` and (suggested) `testEmptyTagHash`)

**Review by oshadmi — COMMENTED** [review]
🤩 Very nice! A few more comments.

File: src/rlookup.h (hunk 11 above)
Comment (raz-mon) [reply src/rlookup.h:75]: Removed as well, since there is no such flag.

File: src/tag_index.c [hunk 21]
```diff
@@ -300,35 +295,7 @@ static TagIndex *openTagKeyDict(RedisSearchCtx *ctx, RedisModuleString *key, int
...
-      }
-    } else {
-      ret = RedisModule_ModuleTypeGetValue(*keyp);
-    }
-  } else {
-    ret = openTagKeyDict(sctx, formattedKey, openWrite);
-  }
-
-  return ret;
+  return openTagKeyDict(sctx, formattedKey, openWrite);
```
[the remark this answers is not in the document]
Comment (raz-mon) [reply src/tag_index.c:298]: If it is tested in our tests - yes.
As we saw the first last version that used this dead code was far before the last version that we currently support.

[the remark this answers is not in the document]
Comment (raz-mon) [reply src/tag_index.c:298]: I don't see any scenario that can reach this currently in our code.

File: src/tokenize.c [hunk 22]
```diff
@@ -95,8 +94,9 @@ uint32_t simpleTokenizer_Next(RSTokenizer *base, Token *t) {
     }
 
     char *normalized = DefaultNormalize(tok, normBuf, &normLen);
-    // ignore tokens that turn into nothing
-    if (normalized == NULL || normLen == 0) {
+
+    // ignore tokens that turn into nothing, unless the whole string is empty.
+    if ((normalized == NULL || normLen == 0) && !empty_input) {
```
[the remark this answers is not in the document]
Comment (raz-mon) [reply src/tokenize.c:99]: Yes, added a condition to the stopword check.
Stemming and phonetics are "automatically" skipped (`normLen` check).

File: src/tokenize.h
[the remark this answers is not in the document]
Comment (raz-mon) [reply src/tokenize.h:1]: No

File: tests/pytests/test_empty.py [hunk 23]
```diff
@@ -0,0 +1,655 @@
...
+    env.expect('FT.CREATE', 'jidx_sortable', 'ON', 'JSON', 'SCHEMA', '$t', 'AS', 't', 'TAG', 'ISEMPTY', 'SORTABLE').ok()
+    EmptyJSONTest(env, 'jidx_sortable')
+    env.flush()
+
+    env.expect('FT.CREATE', 'jidx_suffix', 'ON', 'JSON', 'SCHEMA', '$t', 'AS', 't', 'TAG', 'ISEMPTY', 'WITHSUFFIXTRIE').ok()
+    EmptyJSONTest(env, 'jidx_suffix')
+    env.flush()
+
+    env.expect('FT.CREATE', 'jidx', 'ON', 'JSON', 'SCHEMA', '$arr[*]', 'AS', 'arr', 'TAG', 'ISEMPTY').ok()
+    # Empty array values ["a", "", "c"] with explicit array components indexing
```
[the remark this answers is not in the document]
Comment (raz-mon) [reply tests/pytests/test_empty.py:417]: Elements

File: tests/pytests/test_empty.py (hunk 20 above)
Comment (raz-mon) [reply tests/pytests/test_empty.py:24]: There is a test for empty text with JSON as well, this is the reason I took this funciton outside the tag case.

File: tests/pytests/test_empty.py [hunk 24]
```diff
@@ -0,0 +1,655 @@
...
+            '  }',
+            '}',
+            ''
+        ]
+        env.assertEqual(res, expected)
+
+    # Create an index with a TAG field, that also indexes empty strings, another
+    # TAG field that doesn't index empty values, and a TEXT field
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 't', 'TAG', 'ISEMPTY', 'text', 'TEXT').ok()
+    testEmptyTextHash('idx')
```
[the remark this answers is not in the document]
Comment (raz-mon) [reply tests/pytests/test_empty.py:387]: ?

File: tests/pytests/test_empty.py (hunk 14 above)
Comment (raz-mon) [reply tests/pytests/test_empty.py:232]: Yes

File: tests/pytests/test_empty.py (hunk 17 above)
Comment (raz-mon) [reply tests/pytests/test_empty.py:495]: I would rather it be clear that it was added to the parser here.

File: tests/pytests/test_empty.py (hunk 18 above)
Comment (raz-mon) [reply tests/pytests/test_empty.py:553]: A little bit. But not too much since it doesn't hurt to show that this prefix search doesn't yield empty results.

File: tests/pytests/test_empty.py (hunk 19 above)
Comment (raz-mon) [reply tests/pytests/test_empty.py:137]: Answered above

File: src/document.c [hunk 25]
```diff
@@ -446,7 +447,7 @@ FIELD_PREPROCESSOR(fulltextPreprocessor) {
 
       Token tok = {0};
       while (0 != aCtx->tokenizer->Next(aCtx->tokenizer, &tok)) {
-        if ((!strcmp(tok.tok, "")) && !FieldSpec_IndexesEmpty(fs)) {
+        if (strlen(tok.tok) == 0 && !indexesEmpty) {
```
Comment (oshadmi) [inline src/document.c:450]:
```suggestion
        if (!indexesEmpty && strlen(tok.tok) == 0) {
```

File: src/document.c [hunk 26]
```diff
@@ -688,7 +689,7 @@ FIELD_PREPROCESSOR(geoPreprocessor) {
       break;
     case FLD_VAR_T_BLOB_ARRAY:
     case FLD_VAR_T_NUM:
-    case FLD_VAR_T_GEOMETRY:
+      RS_LOG_ASSERT(0, "Unsupported field type for GEO index");
       RS_LOG_ASSERT(0, "Unsupported field type GEOMETRIY for GEO index");
```
Comment (oshadmi) [inline src/document.c:693]: Redundant logging

**Review by oshadmi — COMMENTED** [review]
⛏️  Minor comments

File: src/document.c (hunk 25 above)
Comment (raz-mon) [reply src/document.c:450]: Assuming that there will be more cases that don't index empty than do (which seems logic at this point).. I can go with this.

**Review by oshadmi — APPROVED** [review]
